### PR TITLE
[JUJU-1057] Fix GitHub Actions on `develop` branch

### PR DESCRIPTION
**Pull request by barrettj12**
[Smoke](https://github.com/juju/juju/runs/6224161870?check_suite_focus=true#step:8:13) and [Snapcraft](https://github.com/juju/juju/runs/6224161450?check_suite_focus=true#step:8:8) tests are returning a strange error:
```
ERROR cannot load ssh client keys: mkdir /home/runner/.local/share/juju: permission denied
```

### CODE REVIEW

File: environs/tools/build.go [hunk 1]
```diff
@@ -246,14 +246,25 @@ func buildJujus(dir string) error {
 	var modArg string
 	var lastErr error
 	for _, m := range []string{"-mod=vendor", "-mod=readonly"} {
+		cmd1 := exec.Command("w")
```
Comment (tlm) [inline environs/tools/build.go:249]: What's this??

File: environs/tools/build.go [hunk 2]
```diff
@@ -246,14 +246,25 @@ func buildJujus(dir string) error {
...
+		cmd2 := exec.Command("echo", "$PATH")
+		cmd2.Env = append(os.Environ(), "GO111MODULE=on")
+		out2, err2 := cmd2.CombinedOutput()
 		cmd := exec.Command("go", "list", m, "github.com/juju/juju")
 		cmd.Env = append(os.Environ(), "GO111MODULE=on")
 		out, err := cmd.CombinedOutput()
 		if err != nil {
 			info := `cannot build juju agent outside of github.com/juju/juju tree
 	cd into the directory containing juju %s %s
+	%s
```
Comment (tlm) [inline environs/tools/build.go:261]: What's this??

File: environs/tools/build.go [hunk 3]
```diff
@@ -246,14 +246,25 @@ func buildJujus(dir string) error {
...
+	%s
+	$ w --> %s
+	%s
+	$ PATH = %s
 	%s`
-			lastErr = errors.Annotatef(err, info, jujuversion.Current.String(), jujuversion.GitCommit, out)
+			lastErr = errors.Annotatef(err, info, jujuversion.Current.String(), jujuversion.GitCommit, out, "DEBUG:", out1, err1, out2, err2)
```
Comment (tlm) [inline environs/tools/build.go:267]: ??

File: .github/workflows/smoke.yml [hunk 4]
```diff
@@ -51,7 +51,11 @@ jobs:
     - name: Smoke Test
       shell: bash
       run: |
-        cd tests && ./main.sh smoke
+        cd tests
```
Comment (tlm) [inline .github/workflows/smoke.yml:54]: Honestly I think we can do this better so we're not left holding the bag with Github changing stuff.

My proposal is to make a new directory in the tests and export the env variable JUJU_DATA to that dir.

We then don't need to rely on .local/share

File: .github/workflows/snap.yml [hunk 5]
```diff
@@ -54,4 +54,7 @@ jobs:
       run: |
         set -euxo pipefail
         lxc network set lxdbr0 ipv6.address none
+        stat /home/runner/.local/share
```
Comment (tlm) [inline .github/workflows/snap.yml:57]: Comments as per above

File: .github/workflows/smoke.yml (hunk 4 above)
Comment (barrettj12) [reply .github/workflows/smoke.yml:54]: Yep I think setting JUJU_DATA is a good idea, to a directory that we can control.

Comment (barrettj12) [reply .github/workflows/smoke.yml:56]: So I set JUJU_DATA and now I'm [getting](https://github.com/juju/juju/runs/6317983150?check_suite_focus=true#step:8:13)
```
ERROR Permission denied, are you in the lxd group?
```
Definitely some kind of weird permissions problem

Comment (barrettj12) [reply .github/workflows/smoke.yml:55]: Tried to setup LXD as per [here](https://juju.is/docs/olm/lxd#heading--configure-lxd), let's see if it helps.

Comment (tlm) [reply .github/workflows/smoke.yml:70]: yeah that's a different problem now. You may want to look at `mktemp -d` for creating the JUJU_DATA directory.

**Comment by barrettj12** [comment]
All the failures seem to be coming back to permission errors. I've also noticed that the failures all come after building a Juju snap, installing it, and trying to use it. My theory is that the Juju snap doesn't have enough permissions to do stuff (access go, access ssh keys, connect to lxd, etc).

**Comment by barrettj12** [comment]
I'm also noticing that when we try to install the Juju snap, it gives a warning:
```
Warning: flag --classic ignored for strictly confined snap juju
```
This seems like it could be causing permission errors.

**Comment by barrettj12** [comment]
Using [this action](https://github.com/mxschmitt/action-tmate), I was able to leave the machines running and create a remote ssh session that you can use.

- Snapcraft / linux-amd64 test:
  - Web shell: https://tmate.io/t/vcHcxkQRWah4mfMzPZYjPQag6
  - SSH: `ssh vcHcxkQRWah4mfMzPZYjPQag6@sfo2.tmate.io`
- Smoke / Smoke test:
  - Web shell: https://tmate.io/t/5EzZjzZJ7nZc95srgrknYmLZc
  - SSH: `ssh 5EzZjzZJ7nZc95srgrknYmLZc@nyc1.tmate.io`
- Smoke / Upgrade (latest/stable, localhost) test:
  - Web shell: https://tmate.io/t/39BtTEy8yx8F3gBuSN95zbANN
  - SSH: `ssh 39BtTEy8yx8F3gBuSN95zbANN@sfo2.tmate.io`

Would appreciate if someone else can take a look at this, as I've definitely reached the limits of my Juju/lxd/snap/linux knowledge.

**Comment by barrettj12** [comment]
Using snap's `--devmode` flag fixed most of the permission errors. Snap seemed to be ignoring the `--classic` flag when installing the Juju snap, which meant it couldn't e.g. bootstrap new controllers.

The smoke tests were also running into versioning issues. I've already reported the bug [here](https://bugs.launchpad.net/juju/+bug/1971909) - the fix I've employed is changing the `juju_version()` function, but maybe we should think about whether this calls for a Juju change.

**Comment by barrettj12** [comment]
**Update on Smoke / Upgrade tests:** part of the workflow is to build the Juju 3.0 snap, install it, then try to upgrade an old controller:
```
juju upgrade-controller -v --build-agent
```
With the --build-agent flag, it's trying to use Go to rebuild the agent binary. However, the $PATH seen by the Juju 3.0 snap is different to the system $PATH. Therefore, Juju cannot find Go and rebuild the binary, which is causing the test to fail.

**Comment by barrettj12** [comment]
# BLOCKED

There are some changes in #14015 that need to be merged through to develop before this PR can land.

Once that happens, I need to remember to change $JUJU_OLDV to `2.9/stable` in the Smoke/Upgrade test.

**Comment by barrettj12** [comment]
I did `snap connections juju` after installing the new Juju snap, to see which connections were working.

In the Smoke and Snapcraft tests, the output is:
```
Interface        Plug                       Slot           Notes
content          juju:peers                 -              -
desktop          juju:desktop               :desktop       -
home             juju:home                  :home          -
lxd              juju:lxd                   -              -
network          juju:network               :network       -
network-bind     juju:network-bind          :network-bind  -
personal-files   juju:config-lxd            -              -
personal-files   juju:dot-aws               -              -
personal-files   juju:dot-azure             -              -
personal-files   juju:dot-google            -              -
personal-files   juju:dot-kubernetes        -              -
personal-files   juju:dot-local-share-juju  -              -
personal-files   juju:dot-maas              -              -
personal-files   juju:dot-openstack         -              -
personal-files   juju:dot-oracle            -              -
ssh-public-keys  juju:ssh-public-keys       -              -
```
I believe slots with `-` are disconnected, which would explain why it's struggling to connect to lxd and the ssh keys.

For the Upgrade tests, the connections seem to be fine.

**Comment by barrettj12** [comment]
In a local lxd container, I tried installing the Juju 3 snap and manually granting permissions with `snap connect`, but it didn't seem to help. I continued to get the error
```
ERROR cannot load ssh client keys: mkdir /home/runner/.local/share/juju: permission denied
```

**Comment by barrettj12** [comment]
Ok, I manually `snap connect`ed things, and now the Snapcraft and Smoke tests are passing. The Upgrade tests are broken due to a [known issue](https://bugs.launchpad.net/juju/+bug/1812191), and the macOS client tests are a separate bag of worms. I'll tidy up and merge.

**Review by wallyworld — APPROVED** [review]

**Pull request merged**